coalib/output/ConsoleInteraction.py

```diff
@@ -131,36 +130,61 @@ def acquire_actions_and_apply(console_printer,
     :param result:          A derivative of Result.
     :param file_dict:       A dictionary containing all files with filename as
                             key.
-    :param cli_actions:     The list of cli actions available.
     """
-    cli_actions = CLI_ACTIONS if cli_actions is None else cli_actions
     failed_actions = set()
     while True:
-        actions = []
-        for action in cli_actions:
-            if action.is_applicable(result, file_dict, file_diff_dict) is True:
-                actions.append(action)
+        actions = result.get_filtered_actions(file_dict, file_diff_dict)
 
         if actions == []:
             return
 
         action_dict = {}
         metadata_list = []
+        # FIXME: If we want to allow users to add multiple actions of same
+        # type, ``metadata.name`` should not be used as a key for action_dict
         for action in actions:
             metadata = action.get_metadata()
             action_dict[metadata.name] = action
             metadata_list.append(metadata)
 
+        applied_action = ask_for_action_and_apply(console_printer,
+                                                  section,
+                                                  metadata_list,
+                                                  action_dict,
+                                                  failed_actions,
+                                                  result,
+                                                  file_diff_dict,
+                                                  file_dict)
+
         # User can always choose no action which is guaranteed to succeed
-        if not ask_for_action_and_apply(console_printer,
-                                        section,
-                                        metadata_list,
-                                        action_dict,
-                                        failed_actions,
-                                        result,
-                                        file_diff_dict,
-                                        file_dict):
+        if not applied_action:
             break
+        else:
+            next_actions = applied_action.next_actions
+
+            # FIXME: currently the code doensn't check for
+            # circular dependencies among the next_actions
```

Comment:
actions aren't dependencies right? And the user decides which to execute next, they may be circular because there's always a way out for the user right?